### PR TITLE
Add edge detection for mask image

### PR DESCRIPTION
**Pull request by wuyuanhang94**
### Description
Add edge detection for mask image

### Status
**Ready**

### Types of changes

- [x] New feature
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: monai/transforms/post/contour.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+import numpy as np
+
+class Contour:
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:15]: 1. Could you please move this transform into `monai.transforms.post.array.py` and follow the same pattern as other transforms?
2. Maybe change the class name to `KeepContour` or `GetContour`?

Thanks.

File: monai/transforms/post/contour.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    """
+    Return the contour of the organ in a mask image that only composes of 0 and 255 in gray value,
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:17]: should suppose values of [0, 1] instead of [0, 255].

File: monai/transforms/post/contour.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        with shape being (1, 1, width, height)
+        '''
+        if self.kernel_type != 'Laplace':
+            raise NotImplementedError
+
+        padding = 0 if self.to_resize else 1
+        conv = nn.Conv2d(1, 1, kernel_size=3, stride=1, padding=padding, bias=False)
+
+        kernel = np.array([[-1, -1, -1], [-1, 8, -1], [-1, -1, -1]], dtype='float32')
+        kernel = torch.from_numpy(kernel.reshape((1, 1, 3, 3)))
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:42]: I would suggest defining kernel by `torch.tensor(XXX, dtype=XXX, device=image.device)` directly.

File: monai/transforms/post/contour.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        conv.weight = torch.nn.Parameter(kernel)
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:43]: should be: `conv.weight = torch.nn.Parameter(kernel, requires_grad=False)` to disable auto grad.

File: monai/transforms/post/contour.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    Args:
+        kernel_type: the method applied to do edge detection.
+        to_resize: whether to keep original shape of `mask_image`
+    """
+
+    def __init__(self, kernel_type='Laplace', to_resize=False):
+        self.kernel_type = kernel_type
+        self.to_resize = to_resize
+
+    def __call__(self, mask_image):
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:29]: maybe use `img` instead of `mask_image`?

File: monai/transforms/post/contour.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            raise NotImplementedError
+
+        padding = 0 if self.to_resize else 1
+        conv = nn.Conv2d(1, 1, kernel_size=3, stride=1, padding=padding, bias=False)
+
+        kernel = np.array([[-1, -1, -1], [-1, 8, -1], [-1, -1, -1]], dtype='float32')
+        kernel = torch.from_numpy(kernel.reshape((1, 1, 3, 3)))
+        conv.weight = torch.nn.Parameter(kernel)
+
+        contour_img = conv(mask_image).data.squeeze().numpy()
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:45]: 1. why `squeeze` here?
2. please return Tensor directly instead of `numpy`.

File: tests/test_contour.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import unittest
+import torch
+from monai.transforms import Contour
+
+def genImg(width, height):
+    img = torch.zeros(width, height, dtype=torch.float)
+    row_idx = torch.randint(0, width, (width // 2, ))
+    col_idx = torch.randint(0, height, (height // 2, ))
+    img[row_idx, col_idx] = 255
```
Comment (Nic-Ma) [inline tests/test_contour.py:20]: please use 1 instead of 255 here.

File: tests/test_contour.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    return img.view(1, 1, width, height)
+    
+class TestContour(unittest.TestCase):
+    def test_contour(self):
+        width, height = 512, 512
+        test_image = genImg(width, height)
+
+        input_param = {"to_resize": False, "kernel_type": "Laplace"}
+        test_result = Contour(**input_param)(test_image)
+        self.assertEqual(test_result.shape, (width, height))
```
Comment (Nic-Ma) [inline tests/test_contour.py:30]: please also add a test for small input data and assert the output values as expected.

File: monai/transforms/post/contour.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    def __call__(self, mask_image):
+        '''
+        Args:
+        mask_image: torch tensor of the mask_image that you want to find the contour of,
+        with shape being (1, 1, width, height)
+        '''
+        if self.kernel_type != 'Laplace':
+            raise NotImplementedError
+
+        padding = 0 if self.to_resize else 1
```
Comment (Nic-Ma) [inline monai/transforms/post/contour.py:38]: I would suggest removing arg `to_resize` and always keep the original size, what do you think?

**Comment by Nic-Ma** [comment]
Hi @wuyuanhang94 ,

BTW, please add support to define the `width` of the edge.
The border curve should be within the segmentation area and the user can set the width of the border curve.

Thanks.

File: monai/transforms/post/contour.py (hunk 1 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: 1. sure, already done
2. I don't think `GetContour` is a good class name, it's more like method name.

File: monai/transforms/post/contour.py (hunk 2 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: I prefer [0, 255], because they respectively represent the gray value of black and white. and the form accords with image binarization.

File: monai/transforms/post/contour.py (hunk 3 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: sure, thanks

File: monai/transforms/post/contour.py (hunk 4 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: sure, thanks

File: monai/transforms/post/contour.py (hunk 5 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: okay

File: monai/transforms/post/contour.py (hunk 6 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: 1. we suppose the batch_size and channel num is 1, squeeze them.
2. sure, done

File: monai/transforms/post/contour.py (hunk 9 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: makes sense, other than original shape can be kept, it can also identically map the index to original image, thanks!

**Comment by wuyuanhang94** [comment]
Hi @Nic-Ma, I don't fully understand why we have the concern, as the output image conved by laplace kernel is still a gray image, I think we cannot interfere or set the width of edge.

File: tests/test_contour.py (hunk 8 above)
Comment (wuyuanhang94) [reply tests/test_contour.py:31]: I think there's no need, as output_width is always equal to input_width when padding and stride are 1.

File: monai/transforms/post/contour.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/post/contour.py]: OK, let's keep the name.

File: monai/transforms/post/contour.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/transforms/post/contour.py]: This post transform is used for `model output` instead of normal 2D images, should not be 255. And even for normal mages, `255` is not commonly used in medical images.
So I think you can suppose input data is within [0, 1] values.

File: monai/transforms/post/contour.py (hunk 6 above)
Comment (Nic-Ma) [reply monai/transforms/post/contour.py]: Actually, here we should suppose `batch = N` and `channel = M`, model output is [N, C, H W, [D]] shape.

File: tests/test_contour.py (hunk 8 above)
Comment (Nic-Ma) [reply tests/test_contour.py:31]: Sorry, maybe I didn't make it clear, I mean to add a test to check all the values in array, for example:
input data:
```
[[0, 0, 0, 1, 1],
[0, 0, 0, 1, 1],
[0, 0, 1, 1, 1],
[0, 1, 1, 1, 1],
[1, 1, 1, 1, 1]]
```
expected output:
```
[[0, 0, 0, 1, 1],
[0, 0, 0, 1, 1],
[0, 0, 1, 0, 1],
[0, 1, 0, 0, 1],
[1, 1, 1, 1, 1]]
```
right?
use `np.testing.assert_allclose(expected, result)` to test.

**Comment by Nic-Ma** [comment]
> Hi @Nic-Ma, I don't fully understand why we have the concern, as the output image conved by laplace kernel is still a gray image, I think we cannot interfere or set the width of edge.

I am OK to make it simple for the first version, may add support to set the width in the future.

File: monai/transforms/post/array.py [hunk 10]
```diff
@@ -248,3 +248,37 @@ def __call__(self, img):
...
+        Returns:
+            A torch tensor representing contour image, note:
+                1. it's gray image, rather than binary classification result of whether edge or not.
+                2. In order to keep the original shape of mask image, we use padding as default.
+        """
+        if self.kernel_type != 'Laplace':
+            raise NotImplementedError
+
+        kernel = torch.tensor([[-1, -1, -1], [-1, 8, -1], [-1, -1, -1]], dtype=torch.float32).view((1, 1, 3, 3), device=img.device)
+        conv = torch.nn.Conv2d(1, 1, kernel_size=3, stride=1, padding=1, bias=False)
```
Comment (Nic-Ma) [inline monai/transforms/post/array.py:279]: Hi @wuyuanhang94 ,

Could you please help support both 2D and 3D input data?
Thanks.

File: monai/transforms/post/contour.py (hunk 2 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: ok, get your point, updated the code, please help review again.

File: monai/transforms/post/contour.py (hunk 6 above)
Comment (wuyuanhang94) [reply monai/transforms/post/contour.py]: ok, already done as you suggested.

File: tests/test_contour.py (hunk 7 above)
Comment (wuyuanhang94) [reply tests/test_contour.py]: ok

File: tests/test_contour.py (hunk 8 above)
Comment (wuyuanhang94) [reply tests/test_contour.py]: thanks Nic, it's really necessary

**Comment by Nic-Ma** [comment]
Hi @wuyuanhang94 ,

Could you please help slightly adjust to fix the code format issues?
/home/runner/work/MONAI/MONAI/monai/transforms/post/array.py:265:10: N802 function name '__findImgContour' should be lowercase
/home/runner/work/MONAI/MONAI/tests/test_contour.py:14:10: E401 multiple imports on one line
/home/runner/work/MONAI/MONAI/tests/test_contour.py:18:2: N816 variable 'expectedOutputForCube' in global scope should not be mixedCase
/home/runner/work/MONAI/MONAI/tests/test_contour.py:104:6: N802 function name 'genFixedCube' should be lowercase
/home/runner/work/MONAI/MONAI/tests/test_contour.py:117:6: N802 function name 'genFixedImg' should be lowercase
/home/runner/work/MONAI/MONAI/tests/test_contour.py:130:6: N806 variable 'expectedOutputForImg' in function should be lowercase
/home/runner/work/MONAI/MONAI/tests/test_contour.py:148:10: N806 variable 'expectedOutput' in function should be lowercase
/home/runner/work/MONAI/MONAI/tests/test_contour.py:159:10: N806 variable 'expectedOutput' in function should be lowercase

Thanks.

**Comment by wuyuanhang94** [comment]
> Hi @wuyuanhang94 ,
> 
> Could you please help slightly adjust to fix the code format issues?
> /home/runner/work/MONAI/MONAI/monai/transforms/post/array.py:265:10: N802 function name '__findImgContour' should be lowercase
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:14:10: E401 multiple imports on one line
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:18:2: N816 variable 'expectedOutputForCube' in global scope should not be mixedCase
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:104:6: N802 function name 'genFixedCube' should be lowercase
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:117:6: N802 function name 'genFixedImg' should be lowercase
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:130:6: N806 variable 'expectedOutputForImg' in function should be lowercase
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:148:10: N806 variable 'expectedOutput' in function should be lowercase
> /home/runner/work/MONAI/MONAI/tests/test_contour.py:159:10: N806 variable 'expectedOutput' in function should be lowercase
> 
> Thanks.
@Nic-Ma modified, please check again then promotion, thanks!

**Comment by Nic-Ma** [comment]
Hi @wyli , could you please help double confirm the algorithm in this PR?
Thanks.

File: monai/transforms/post/array.py [hunk 11]
```diff
@@ -248,3 +248,60 @@ def __call__(self, img):
...
+                3. The edge detection is just approximate due to
+                    a) defects inherent to Laplace kernel, ideally the edge should be thin enough, but now it has a thickness.
+                    b) need to search the optimal/better thresold for classification
+        """
+        if self.kernel_type != "Laplace":
+            raise NotImplementedError
+        if img.ndim != 4 and img.ndim != 5:
+            raise RuntimeError("img.ndim should be 4 or 5")
+        if img.ndim == 4:
+            return self.__find_img_contour(img)
```
Comment (wyli) [inline monai/transforms/post/array.py:295]: why is the 2D case implemented as a separate private method?

File: monai/transforms/post/array.py [hunk 12]
```diff
@@ -248,3 +248,60 @@ def __call__(self, img):
             output = img
 
         return output
+
+
+class Contour(Transform):
```
Comment (wyli) [inline monai/transforms/post/array.py:253]: I feel the design needs to be improved, please see kornia's abstraction of building kernels, setting up the padding, etc https://github.com/kornia/kornia/tree/master/kornia/filters. also the gaussian filtering in monai shows a possibility of handling 1d/2d/3d: https://github.com/Project-MONAI/MONAI/blob/master/monai/networks/layers/simplelayers.py#L79

Comment (Nic-Ma) [reply monai/transforms/post/array.py]: good idea!

File: monai/transforms/post/array.py (hunk 11 above)
Comment (Nic-Ma) [reply monai/transforms/post/array.py:339]: Will refactor in my PR together.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I saw you marked "request change", could you please help review it again?
Thanks.

File: monai/transforms/post/array.py (hunk 11 above)
Comment (wuyuanhang94) [reply monai/transforms/post/array.py:339]: because we are using different kernels for 2D and 3D

File: monai/transforms/post/array.py (hunk 12 above)
Comment (wuyuanhang94) [reply monai/transforms/post/array.py]: thanks for suggestions, we can create a separate `build_kernel` method for laplace kernel on different dimensions in the near future; padding equal to 1 has been fixed for this time being.
